applications/FluidDynamicsApplication/custom_elements/fluid_adjoint_element.h

```diff
@@ -0,0 +1,325 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:         BSD License
+//                   Kratos default license: kratos/license.txt
+//
+//  Main authors:    Suneth Warnakulasuriya
+//
+
+#if !defined(KRATOS_FLUID_ADJOINT_ELEMENT_H)
+#define KRATOS_FLUID_ADJOINT_ELEMENT_H
+
+// System includes
+
+// External includes
+
+// Project includes
+#include "includes/checks.h"
+#include "includes/constitutive_law.h"
+#include "includes/element.h"
+#include "includes/properties.h"
+#include "utilities/adjoint_extensions.h"
+
+// Application includes
+
+namespace Kratos
+{
+///@name Kratos Classes
+///@{
+
+template <unsigned int TDim, unsigned int TNumNodes, class TAdjointElementData>
+class FluidAdjointElement : public Element
```

Comment:
Why not deriving from the `FluidElement`?

Reply:
Because `TAdjointElementData` is different from `TElementData`. I was not able to use same `TElementData` from fluid elements because structure is different.

Reply:
But can't the `TAdjointElementData` be an extension of `TEelementData` as all the elements do?

Reply:
I was thinking of doing that when I was first writing this, I don't remember all the points. But followings are few points which I remember.
1. `FluidElementData::UpdateGeometryValues`, `FluidElementData::Initialize` are virtual methods. I wanted to reduce as much as possible vtable lookups since everything in adjoints can be determined at compiler time (And they are called many time in adjoint sovle when constructing a local element matrix than the primal solve). I wanted to take advantage of that fact.
2. `FillFromHistoricalNodalData` methods do loop over nodes in geometry for each different variable. I wanted to use `FluidCalculationUtilities::EvaluateInPoint` and `FluidCalculationUtilities::EvaluateGradientInPoint` which was templated, and again we know what those are at compiler time
3. `FluidElementData` already has nested templated argument, if I'm to use this, then I will end up nesting it again which I think will be harder to debug, and harder to maintain as well. [ `FluidElementData` and its nested versions are complicated enough ]

The third point being the killer, I thought of going for seperated `TAdjointElementData`. I'm in favor of keeping them seperate, we have tests to identify if something from primal side changes to correctly do changes in the adjoint side.